### ISSUE TITLE
Delta time var passed in as function every other frame

### ISSUE THREAD

**Issue by falsechicken** [bakpakin/tiny-ecs#11]
Hello! I am playing around with Love2D and decided to use Tiny-ecs with my project. All was going ok until I tried to use the delta time var. I try to use it for my player movement but every other frame it is a function for some reason. 

My code:

`world = {}`
`world = tiny.world(inputSystem, renderSystem, movementSystem, entity, entity2 )`

`function love.update(_dt)`
    `world:update(_dt)`
`end`

`function InputSystem:process(_e, _dt)`
    `print(_dt)`
`end`

-- This code results in --
function: 0x403c4c20
0.016999974999635
function: 0x403c4c20
0.01636593999865
function: 0x403c4c20
0.024445032999211
function: 0x403c4c20
0.0088829689993872
function: 0x403c4c20
0.018819653996616
function: 0x403c4c20
0.014247577004426
function: 0x403c4c20

So when I try to move my player multiplying by delta time I get an error every other frame because I am trying to multiply by a function. Am I doing something wrong? Thanks!

Actual Full Source: [http://pastebin.com/zkQQfA0v](http://pastebin.com/zkQQfA0v)

**Comment by bakpakin** [comment]
It looks to me that the issue is in love.draw. You pass in the filter in place of dt, which is why it spits out a function rather than dt every other update (you are update twice a frame because the filter is in the wrong place). You should explicitly set dt to 0 in love.draw.
Change love.draw to

``` lua
function love.draw()
    world:update(0, renderSystem.filter)
end
```

**Comment by falsechicken** [comment]
Wow that was it! Thanks! This is a nice little library :)

**Comment by falsechicken** [comment]
So I have just discovered interesting. Passing in zero for the delta in the render loop prevents anything from being rendered. If I omit the zero I cannot press the keys because it would crash as before. But the text renders to the screen. If I pass 0 the text does not render even though the loop is running as indicated by my print statements :/. Any suggestions? Thanks for taking the time to answer these!

EDIT: I seem to be doing filters wrong. Its actually if I omit the filter that the text shows. But that kinda makes little sense as I said before when using the filter the render system still ran. It printed the messages that it was running but wouldn't draw.

EDIT2: Something is defiantly up. All the data it needs to draw should be there. I added some extra print statements to the render system and all the data checks out. The graphics.print function just is not working. Ive been hacking with this for a while now and I am at a loss :/. I don't get it. It doesn't make sense.

-- Console Output --
Render Called.
entity 1
50 50
0.016027238001698
Render Called.
entity 2
100 100
0.016027238001698

My render code: [http://pastebin.com/aK2a67gD](http://pastebin.com/aK2a67gD)

**Comment by bakpakin** [comment]
I have been looking at your code, and you seem confused about how the filter function works when passed to world.update, perhaps because the filter concept is being reused in a slightly different manner here. When passed to world.update, filters are applied to each system to check if they should be run or not, not to entities.

Reusing renderSystem.filter in love.draw is not what you want to do. Want you want is to somehow mark each system that is a 'render system', and filter each system based on that marking. When you create the system, you should just add a flag `isRenderSystem` to it.

``` lua
-- Wherever you create your rendering system
local renderSystem = tiny.processingSystem()
renderSystem.isRenderSystem = true

function love.update(dt)
    world.update(dt)
end
-- Should become
function love.update(dt)
    world.update(dt, tiny.rejectAny('isRenderSystem')
end


function love.draw()
    world.update(0, renderSystem.filter)
end
-- Should become
function love.draw()
    world.update(0, tiny.requireAll('isRenderSystem'))
end
```

In hindsight, I think using filters for choosing different systems is a misleading and confusing API.
I don't think you are the first person with this issue, and I am considering introducing a friendlier concept for updating specific groups of systems.

**Comment by falsechicken** [comment]
Ahhhh.... I see. Yes that is different than how I imagined the filters working. So by flag you mean create a boolean variable in the RenderSystem called isRenderSystem? Or is a "flag" something tiny specific?

EDIT: Creating that bool seemed to work. Just set RenderSystem.isRenderSystem to true causes the render to function. Awesome! Thanks! Yeah it is a bit confusing xD. But with your help I was able to solve so issue so thanks again!
